kuri_camera/src/ros_publisher.cpp

```diff
@@ -43,44 +42,42 @@ struct KuriCameraROSPublisher {
 
   void listen_publish_loop() {
     ros::Rate sub_poll(1);
+    ros::Rate fps(20);
+    cv::Mat image;
+    ros::Time timeForMsg;
+    //cv::Mat imageMat;
     while (ros::ok()) {
       ros::spinOnce();
-      if (publisher.getNumSubscribers() == 0) {
-        ROS_INFO("No subs");
+      fps.sleep();
+      if (imagePublisher.getNumSubscribers() == 0) {
+        ROS_INFO("Waiting for subscribers");
         sub_poll.sleep();
-        //continue;
-      }
-      ROS_INFO("Heard sub");
-      // Someone is subscribed! Get a frame
-      {
-        std::unique_lock<std::mutex> lock(m);
-        frame_processed = false;
-        got_frame.wait(lock, [this]() { return frame_processed; });
-        ROS_ERROR("LOOP SEES PROCESSED");
+        continue;
       }
 
-
+      std::unique_lock<std::mutex> imageLock(imageMutex);
+      if (hasNewImage) {
+        hasNewImage = false;
+        image = cv::imdecode(imageMat, CV_LOAD_IMAGE_UNCHANGED);
+        timeForMsg = recvTime;
+        imageLock.unlock();
+        sensor_msgs::CompressedImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", image).toCompressedImageMsg();
+        msg->header.stamp = timeForMsg;
+        msg->header.frame_id = "upward_looking_camera";
+        imagePublisher.publish(msg);
+      } else {
+        imageLock.unlock();
+      }
     }
   }
 
   void data_callback(uint8_t *buffer, uint32_t size) {
-    std::unique_lock<std::mutex> lock(m);
-    std::cout << size << std::endl;
-    std::cout << (void*)buffer << std::endl;
-    std::cout << buffer[0] << std::endl;
-
-    if (size < 1000) {
-      return;
-    }
-    cv::Mat image_nv12(640, 480, CV_8UC1, buffer);
-    cv::Mat image_bgr;
-    cv::cvtColor(image_nv12, image_bgr, CV_YUV2BGR_YV12);
-    sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", image_bgr).toImageMsg();
-    publisher.publish(msg, camera_info);
-    frame_processed = true;
-    lock.unlock();
-    got_frame.notify_all();
-    ROS_ERROR("CALLBACK RETURNING");
+    if (size < 1000) return;
```

Comment:
May want to add a comment about this. If I recall it's because some weird packets come through when the socket is first started?

Reply:
I removed it -- I think this was a legacy of the linking problems